### DISCUSSION TITLE
How to run spine 3.8 animation using axmol v2.0.0

### DISCUSSION THREAD

**Discussion by nirobhasanbd** [axmolengine/axmol#1470] [General]
As the default spine version is 4.1 so I can not run my spine v3.8 animation. But in change-log they stated that supported spine version is 3.6~4.1, default is 4.1.
So how can I run spine v3.8 animation.
Please help. I am stuck at here.

**Comment by rh101** [comment]
> As the default spine version is 4.1 so I can not run my spine v3.8 animation. But in change-log they stated that supported spine version is 3.6~4.1, default is 4.1. So how can I run spine v3.8 animation. Please help. I am stuck at here.

For v1 of Axmol, you can just replace the folder in Axmol (`axmol/extensions/spine`) with the one in this repo (branch 3.8):
https://github.com/axmolengine/spine-axmol/

For v2, it requires some modification, which has been done in this PR (works with both v1 and v2): 
https://github.com/axmolengine/spine-axmol/pull/1

You can also disable the Spine extension in Axmol, add the branch 3.8 folder to your own project, and link it in your project `CMakeLists.txt`.

**Reply by nirobhasanbd** [reply]
@rh101 
Thanks for you quick and details response.

Regarding
> For v1 of Axmol, you can just replace the folder in Axmol (axmol/extensions/spine) with the one in this repo (branch 3.8):
https://github.com/axmolengine/spine-axmol/

This is not working. If I replace the spine folder from https://github.com/axmolengine/spine-axmol/ (`branch 3.8`) then there are some error about `File not found` when I create a new project and run it.

Regarding
> For v2, it requires some modification, which has been done in this PR (works with both v1 and v2):
https://github.com/axmolengine/spine-axmol/pull/1

This is also not working. If I replace the spine folder from https://github.com/axmolengine/spine-axmol/ (`branch 3.8` as the giver PR is merged into `branch 3.8`) then the app is crashing when I create a new project and run it.

Here is my spine files created by `v3.8.75`. Please check the attached files.
[skeletal_spine3.8.75.zip](https://github.com/axmolengine/axmol/files/13521173/skeletal_spine3.8.75.zip)

**Reply by rh101** [reply]
> This is not working. If I replace the spine folder from https://github.com/axmolengine/spine-axmol/ (`branch 3.8`) then there are some error about `File not found` when I create a new project and run it.

I can't help if I don't see the error you're getting.  Show what error it is, and also screenshot the contents of the `extensions/spine` folder.

The spine folder that is in Axmol 2 looks like this:
![image](https://github.com/axmolengine/axmol/assets/8603230/5397ff78-ec2f-48cb-a9eb-ecd0361f604d)

When you copy over the contents of branch 3.8 from the [repo](https://github.com/axmolengine/spine-axmol/tree/spine38), it should look like this:
![image](https://github.com/axmolengine/axmol/assets/8603230/7fec4575-af9e-4eab-b022-7ac5fe5b8be9)

Are you sure you replaced the existing "spine" folder with the correct version?


> This is also not working. If I replace the spine folder from https://github.com/axmolengine/spine-axmol/ (`branch 3.8` as the giver PR is merged into `branch 3.8`) then the app is crashing when I create a new project and run it.

Once again, stating that "the app is crashing" is of no help whatsoever, and you need to copy the crash log and paste it in your post.

**Reply by nirobhasanbd** [reply]
@rh101 
Thanks for the quick response.
Here is my development environment:
`MacBook Pro M2 2022, macOS Sonoma 14.0, Xcode 15.0, Spine 3.8.75`

**Regarding Axmol 2.0 (dev branch)**
I replace the `spine` folder with `spine38` and then create a new project and add spine code like below:
```
string folderReplaceSkeletalSpine = "SkeletalAnimation/";
int productId = 9876;
string strProduct = "skeleton";
string sequence = "idle";
string _strFolderLocation=folderReplaceSkeletalSpine+to_string(productId)+"/";

spine::SkeletonAnimation *skeletonNode = spine::SkeletonAnimation::createWithJsonFile(_strFolderLocation+strProduct+".json", _strFolderLocation+strProduct+".atlas");
skeletonNode->setScaleX(-1.2f);
skeletonNode->setScaleY(1.2f);
skeletonNode->setAnimation(0, sequence, true);
skeletonNode->setPosition(Point(visibleSize.width/2, 120.0f));
this->addChild(skeletonNode, 10);
```
And the path is correct and added the spine resources correctly.
But when I run the project in my `iPhone 12 iOS 17.1` device then the below crash occur.

![Screenshot 2023-12-01 at 12 28 45 PM](https://github.com/axmolengine/axmol/assets/62602617/1e27b677-6a89-4eff-be47-ed059f2240c6)

My spine folder looks like below:

![Screenshot 2023-12-01 at 12 44 53 PM](https://github.com/axmolengine/axmol/assets/62602617/3494aa3f-dcde-48c4-aebd-cd611f9c47a3)

**Reply by rh101** [reply]
The error log is showing you exactly what the problem is.  Look closer:

![image](https://github.com/axmolengine/axmol/assets/8603230/8f9925f1-3eab-4e96-b3f1-ce52d9936407)

If you search the code for the error text, "Unsupported skeleton data, please export with a newer version of Spine.", you would have found this bit of code in `SkeletonJson::readSkeletonData()`:

![image](https://github.com/axmolengine/axmol/assets/8603230/baa36df6-6437-4386-8abb-21fe7b14020d)

The spine file you attached earlier is version 3.8.75:
![image](https://github.com/axmolengine/axmol/assets/8603230/85171922-3d86-4bf3-9cc3-ce2f535c377b)

Spine is working as it should, and the issue is your spine skeleton file. You need to export it with the *latest* version of 3.8 (I think it's 3.8.99 or something similar).

**Reply by nirobhasanbd** [reply]
@rh101 
Thanks.
I can't understand why the minor version does matter. As the spine version is 3.8 so it should support all the minor versions 3.8.0~3.8.99
However thanks a lot to debug the root cause and help me very fast.

**Reply by nirobhasanbd** [reply]
@rh101 
Could you please tell me one thing, why we assert if the spine version is 3.8.75 ? What is the problem with this specific version ?
If I remove this version check condition then what will happen. Does it make any problem ?

![image](https://github.com/axmolengine/axmol/assets/62602617/128a418f-a8c8-4af2-885b-67e3fca408c4)

Sorry for the inconvenience.

**Reply by rh101** [reply]
> Could you please tell me one thing, why we assert if the spine version is 3.8.75 ? What is the problem with this specific version ?
> If I remove this version check condition then what will happen. Does it make any problem ?

I don't have the answers to those questions, and if you look at that section of code, you would notice that it's located in the Spine runtime, so your questions should be directed at the developers of that runtime.  Either post it on their [github](https://github.com/EsotericSoftware/spine-runtimes/) or their [forums](https://esotericsoftware.com/forum/).
